src/method.c

```diff
@@ -41,249 +41,199 @@ static void check_c_types(const char *where, jl_value_t *rt, jl_value_t *at)
+        if (!jl_is_symbol(cc)) {
+            JL_TYPECHK(ccall method definition, tuple, cc);
+            if (jl_nfields(cc) != 2) {
+                jl_error("In ccall calling convention, expected two argument tuple or symbol.");
             }
-            if (e->head == jl_cfunction_sym) {
-                JL_NARGS(cfunction method definition, 5, 5); // (type, func, rt, at, cc)
-                jl_task_t *ct = jl_current_task;
-                jl_value_t *typ = jl_exprarg(e, 0);
-                if (!jl_is_type(typ))
-                    jl_error("first parameter to :cfunction must be a type");
-                if (typ == (jl_value_t*)jl_voidpointer_type) {
-                    jl_value_t *a = jl_exprarg(e, 1);
-                    JL_TYPECHK(cfunction method definition, quotenode, a);
-                    *(jl_value_t**)a = jl_toplevel_eval(module, *(jl_value_t**)a);
-                    jl_gc_wb(a, *(jl_value_t**)a);
-                }
-                jl_value_t *rt = jl_exprarg(e, 2);
-                jl_value_t *at = jl_exprarg(e, 3);
-                if (!jl_is_type(rt)) {
-                    JL_TRY {
-                        rt = jl_interpret_toplevel_expr_in(module, rt, NULL, sparam_vals);
-                    }
-                    JL_CATCH {
-                        if (jl_typetagis(jl_current_exception(ct), jl_errorexception_type))
-                            jl_error("could not evaluate cfunction return type (it might depend on a local variable)");
-                        else
-                            jl_rethrow();
-                    }
-                    jl_exprargset(e, 2, rt);
-                }
-                if (!jl_is_svec(at)) {
-                    JL_TRY {
-                        at = jl_interpret_toplevel_expr_in(module, at, NULL, sparam_vals);
-                    }
-                    JL_CATCH {
-                        if (jl_typetagis(jl_current_exception(ct), jl_errorexception_type))
-                            jl_error("could not evaluate cfunction argument type (it might depend on a local variable)");
-                        else
-                            jl_rethrow();
-                    }
-                    jl_exprargset(e, 3, at);
-                }
-                check_c_types("cfunction method definition", rt, at);
-                JL_TYPECHK(cfunction method definition, quotenode, jl_exprarg(e, 4));
-                JL_TYPECHK(cfunction method definition, symbol, *(jl_value_t**)jl_exprarg(e, 4));
-                return expr;
+            JL_TYPECHK(ccall method definition, symbol, jl_get_nth_field(cc, 0));
+            JL_TYPECHK(ccall method definition, uint16, jl_get_nth_field(cc, 1));
+        }
+        jl_exprargset(e, 0, resolve_definition_effects(jl_exprarg(e, 0), module, sparam_vals, binding_effects, 1));
+    }
+    if (e->head == jl_call_sym && jl_expr_nargs(e) == 3 &&
+            jl_is_globalref(jl_exprarg(e, 0)) &&
```

Comment:
Do we still care about this hack (replacing the syntax `a.b` with `GlobalRef(a, :b)` if `a` appears to be a module?

Reply:
I think that hack has been broken for a long time.

Reply:
okay, yeah, do we just clean this up by deleting it too?
```
julia> @eval Base f() = Base.x
f (generic function with 1 method)

julia> @code_lowered Base.f()
CodeInfo(
    @ REPL[10]:1 within `unknown scope`
1 ─ %1 = Base.Base
│   %2 =   dynamic Base.getproperty(%1, :x)
└──      return %2
)
```

Reply:
Fine by me. We could consider putting it into lowering though.

Reply:
Lowering doesn't know the symbol will refers to a module constant, so it had to be when we evaluate the expression here

Reply:
Alright, fair enough. In that case, I don't think we can really do this at all. With binding redefinition, our semantics do not allow us to make that decision here.